#' Simulate summary statistics using msABC
#' @description This function simulates summary statistics using msABC.
#'              It is optimized for nextgen data, but it should also work for sanger.
#'              This will simulate the entire loci, not a single SNP.
#'              You should include all loci in the model object, including invariable ones. Use the get.data.structure function to set up the parameters of the loci - bp and number of individuals - to be simulated.
#' @param model A model object bult by the main.menu function. Both genomic and sanger-type models are allowed.
#' @param use.alpha Logical. If TRUE the most recent population size change will be exponential. If FALSE sudden demographic changes. Default is FALSE.
#'                  This argument changes ONLY the MOST RECENT demographich change.
#' @param nsim.blocks Number of blocks to simulate. The total number of simulations is: nsim.blocks x sim.block.size.
#' @param sim.block.size Simulations are performed in blocks. This argument defines the size of the block in number of simulations, i.e. how many simulations to run per block.
#'                       A block of 1000 will work for most cases. Increse the total number of simulations with nsim.block argument.
#' @param path Path to write the output. By default outputs will be saved in the working directory.
#' @param output.name String. The prefix of the output names. Defalt is "model"
#' @param append.sims Logical. If TRUE simulations will be appended in the last output. Default is FALSE.
#' @param msABC.call String. Path to the msABC executable. msABC binaries for Mac and Linux are included in the package and should work cases for these operating systems.
#'                   There is no need to change this unless you want to compile the program yourself and point the function to it.
#' @return Writes simulations and parameters to the path directory.
#' @references Hudson R.R. (2002) Generating samples under a Wright-Fisher neutral model of genetic variation. Bioinformatics, 18, 337–338.
#' @references Pavlidis P., Laurent S., & Stephan W. (2010) msABC: A modification of Hudson’s ms to facilitate multi-locus ABC analysis. Molecular Ecology Resources, 10, 723–727.
#' @author Marcelo Gehara
#' @export
sim.msABC.sumstat<-function(model, nsim.blocks, path=getwd(), use.alpha=F, mu.rates=NULL,
                            append.sims=F,block.size=10, msABC.call=get.msABC(),output.name,ncores){

  if(class(model)!="Model"){
    stop("First arguments should be an object of class Model generated by the main.menu() function.")
  }
  if(model$I[1,1]=="genomic"){
    stop("You need to get the data structure first. See function get.data.structure")
  }
  # set working directory
  setwd(path)

  locfile <- get.locfile(model)
  msABC.path <- find.package("PipeMaster")

  if(append.sims==F){
    com<-msABC.commander(model,use.alpha=use.alpha,arg=1)
    write.table(locfile,paste(".",1,"locfile.txt",sep=""),row.names = F,col.names = T,quote = F,sep=" ")
    options(warn=-1)
    x<-strsplit(system2(msABC.call, args=com[[1]], stdout = T,stderr=T,wait=T),"\t")
    options(warn=0)
    nam<-x[1][[1]]
    TD_denom<-paste(nam[grep("pi",nam)],nam[grep("_w",nam)],sep="_")
    #nam<-nam[-grep("ZnS",nam)]
    #nam<-nam[-grep("thomson",nam)]
    nam<-c(nam, TD_denom)
    nam<-c(com[[2]][1,],"mean.rate","sd.rate",nam)
    #t(paste(nam,"_skew",sep="")),t(paste(nam,"_var",sep="")))
    write.table(t(nam),file=paste("SIMS_",output.name,".txt",sep=""),quote=F,row.names = F,col.names = F, append=F,sep="\t")
}

  write(paste('arg <- commandArgs(TRUE)',
              'cat(paste("Core",arg),sep="\\n")',
              "suppressMessages(library(PipeMaster))",
              #"suppressMessages(library(devtools))",
              #"load_all('~/Github/PipeMaster')",
              'load(file=".PM_objects.RData")',
              "res<-sim.func(arg)",
              'write.table(res,file=paste(".",arg,"SIMS_",output.name,".txt",sep=""),quote=F,row.names = F,col.names = F, append=F,sep="\t")',
              'write(1,".log",append=T,sep="\\n")',
              'file.remove(paste(".",arg,"locfile.txt",sep=""))',
              "quit(save='no')",sep="\n"),".script_parallel.R")

  sim.func<-function(arg){

      simulations<-NULL
      for(i in 1:block.size){
        if(!(is.null(mu.rates))){
          rates<-do.call(mu.rates[[1]],args=mu.rates[2:length(mu.rates)])
          rates<-rep(rates, each=as.numeric(model$I[1,3]))
          rates<-list(rates,c(0,0))
        } else {
        rates <- sample.mu.rates(model)
        }
        locfile[,5] <- rates[[1]]
        write.table(locfile,paste(".",arg,"locfile.txt",sep=""),row.names = F,col.names = T,quote = F,sep=" ")
        com <- msABC.commander(model, use.alpha=use.alpha,arg=arg)
        options(warn=-1)
        sumstat <- read.table(text=system2(msABC.call, args=com[[1]], stdout = T,stderr=T,wait=T),header=T,sep="\t")
        options(warn=0)
        sumstat <- subset(sumstat, select=-c(X))
        TD_denom <- data.frame(sumstat[,grep("pi",colnames(sumstat))]-sumstat[,grep("_w",colnames(sumstat))])
        colnames(TD_denom) <- paste(colnames(sumstat)[grep("pi",colnames(sumstat))],sumstat[,grep("_w",colnames(sumstat))])
        #sumstat<-sumstat[,-grep("ZnS",colnames(sumstat))]
        #sumstat<-sumstat[,-grep("thomson",colnames(sumstat))]
        sumstat <- cbind(sumstat, TD_denom)
        #Mean<-apply(sumstat,2,mean, na.rm = T)
        #var<-apply(sumstat,2,var, na.rm=T)
        #kur<-apply(sumstat,2,kurtosis, na.rm=T)
        #skew<-apply(sumstat,2,skewness, na.rm=T)
        param <- c(com[[2]][2,],rates[[2]])
        names(param) <- c(com[[2]][1,],"mean.rate","sd.rate")
      simulations <- rbind(simulations,c(param,sumstat))
      }
    return(simulations)
  }

  parentls <- function()ls(envir=parent.frame())
  save(list=parentls(),file='.PM_objects.RData')

  thou<-0
  for(j in 1:nsim.blocks) {

    write(0,".log")

    for(c in 1:ncores){
      system(paste("Rscript .script_parallel.R",c), wait = F)
    }

    l<-"0"
    TIM1<-system.time(
      while(sum(as.numeric(unlist(strsplit(l, "")))) < ncores){
        Sys.sleep(5)
        l<-readLines(".log")
      })[3]

    #file.remove(".log")

    simulations<-NULL
    print("Reading simulations from slave nodes")
    TIM2<-system.time(
      for(t in 1:ncores){
        SIM <- read.table(paste(".",t,"SIMS_",output.name,".txt",sep=""), sep="\t")
        simulations<-rbind(simulations, SIM)
        })[3]
    print("FINISHED -- Reading simulations from slave nodes")
    TIM3<-system.time(
      print("writing simulations to file"),
      write.table(simulations,file=paste("SIMS_",output.name,".txt",sep=""),quote=F,row.names = F,col.names = F, append=T,sep="\t")
    )[3]


    TIM4<-system.time(
      for(t in 1:ncores){
    file.remove(paste(".",t,"SIMS_",output.name,".txt",sep=""))
      })[3]

    thou<-thou+block.size*ncores
    Total.time<-round(((((TIM1+TIM2+TIM3+TIM4)*nsim.blocks)-((TIM1+TIM2+TIM3)*j))/60)/60,3)
    cat(paste("PipeMaster:: ",thou," (",round(((block.size*ncores)/(TIM1+TIM2+TIM3))*60*60)," sims/h) | ",Total.time," hours remaining",sep=""),"\n")
    #write.table(param,file=paste(output.name,"_par.txt",sep=""),quote=F,row.names = F,col.names = F, append=T,sep="\t")
  }

}

