#' Simulate summary statistics using msABC
#' @param model A model object bult by the main.menu function. Both genomic and sanger type models are allowed.
#' @param use.alpha Logical. If TRUE the most recent population size change will be exponential. If FALSE sudden demographic changes. Default is FALSE.
#'                  This argument changes ONLY the MOST RECENT demographich change.
#' @param nsim.blocks Number of blocks to simulate. The total number of simulations is: nsim.blocks x sim.block.size.
#' @param sim.block.size Simulations are performed in blocks. This argument defines the size of the block in number of simulations, i.e. how many simulations to run per block.
#'                       A block of 1000 will work for most cases. Increse the total number of simulations with nsim.block argument.
#' @param path Path to write the output. By default outputs will be saved in the working directory.
#' @param output.name String. The prefix of the output names. Defalt is "model"
#' @param append.sims Logical. If TRUE simulations will be appended in the last output. Default is FALSE.
#' @param get.moments Logical. If TRUE computes the four moments (mean, variance, kurtosis, skewness) of each summary statistics across loci. If False only mean is computed. Defalt is FALSE.
#' @param msABC.call String. Path to the msABC executable. msABC binaries for Mac and Linux are included in the package and should work cases for these operating systems.
#'                   There is no need to change this unless you want to compile the program yourself and point the function to it.
#' @return Writes simulations and parameters to the path directory.
#' @references Hudson R.R. (2002) Generating samples under a Wright-Fisher neutral model of genetic variation. Bioinformatics, 18, 337–338.
#' @references Pavlidis P., Laurent S., & Stephan W. (2010) msABC: A modification of Hudson’s ms to facilitate multi-locus ABC analysis. Molecular Ecology Resources, 10, 723–727.
#' @author Marcelo Gehara
#' @export
sim.msABC.sumstat<-function(model,nsim.blocks,path=getwd(),use.alpha=F,
                          append.sims=F,block.size=10, msABC.call=get.msABC(),output.name,ncores){

  if(class(model)!="Model"){
    stop("First arguments should be an object of class Model generated by the main.menu() function.")
  }
  # set working directory
  setwd(path)

  if(append.sims==F){
    if(model$I[1,1]=="genomic"){
      com<-PipeMaster:::ms.commander.snp(model,use.alpha=use.alpha,msABC=msABC.call)
      } else {
      com<-PipeMaster:::msABC.commander(model,use.alpha=use.alpha,msABC=msABC.call)
      }

    x<-strsplit(system(com[[1]],intern=T),"\t")
    nam<-x[1][[1]]
    TD_denom<-paste(nam[grep("pi",nam)],nam[grep("theta_w",nam)],sep="_")
    nam<-nam[-grep("ZnS",nam)]
    nam<-nam[-grep("thomson",nam)]
    nam<-c(nam, TD_denom)
    if(model$I[1,1]=="genomic"){
      nam<-c(t(paste(nam,"_mean",sep="")),t(paste(nam,"_var",sep="")))
      nam<-c(com[[3]][1,],nam)
    }else{
    nam<-c(t(com[[length(com)]][1,1:(ncol(com[[length(com)]])-(nrow(model$loci)-2))]),t(paste(nam,"_mean",sep="")),t(paste(nam,"_var",sep="")))
    }
         #t(paste(nam,"_skew",sep="")),t(paste(nam,"_var",sep="")))
    write.table(t(nam),file=paste("SIMS_",output.name,".txt",sep=""),quote=F,row.names = F,col.names = F, append=F,sep="\t")
    }

  write(paste('args <- commandArgs(TRUE)',
              'cat(paste("Core",args),sep="\n")',
            "suppressMessages(library(PipeMaster))",
            'load(file=".PM_objects.RData")',
            "res<-sim.func()",
            'write.table(res,file=paste(".",args,"SIMS_",output.name,".txt",sep=""),quote=F,row.names = F,col.names = F, append=F,sep="\t")',
            'write(1,".log",append=T)',
            "quit(save='no')",sep="\n"),".script_parallel.R")

  sim.func<-function(){

    if(model$I[1,1]=="genomic"){

      simulations <- NULL
      for(i in 1:block.size) {
        com<-PipeMaster:::ms.commander.snp(model,msABC=msABC.call, use.alpha = use.alpha)
        #system(paste(com[[1]]," > ",i,"out.txt",sep=""))
        #sumstat<-read.table(paste0(i,"out.txt"),header = T)
        sumstat<-read.table(text=system(paste(com[[1]],sep=""),intern=T),header=T,sep="\t")
        sumstat<-subset(sumstat,select=-X)

        TD_denom<-data.frame(sumstat[,grep("pi",colnames(sumstat))]-sumstat[,grep("theta_w",colnames(sumstat))])
        colnames(TD_denom)<-paste(colnames(sumstat)[grep("pi",colnames(sumstat))],
                                  colnames(sumstat)[grep("theta_w",colnames(sumstat))],sep="_")
        sumstat<-sumstat[,-grep("ZnS",colnames(sumstat))]
        sumstat<-sumstat[,-grep("thomson",colnames(sumstat))]
        sumstat<-cbind(sumstat, TD_denom)

        Mean<-colMeans(sumstat,na.rm = T)
        var<-apply(sumstat,2,var, na.rm=T)
        #kur<-apply(sumstat,2,kurtosis, na.rm=T)
        #skew<-apply(sumstat,2,skewness, na.rm=T)

        param<-as.numeric(com[[3]][2,])
        names(param)<-com[[3]][1,]
        simulations<-rbind(simulations,c(param,Mean,var))
      }
    } else {
      simulations<-NULL
      #param<-NULL
      for(i in 1:block.size){

        com <- PipeMaster:::msABC.commander(model,use.alpha=use.alpha, msABC = msABC.call)
        sumstat<-NULL
        for(u in 1:nrow(model$loci)){
          x<-read.table(text=system(paste(com[[u]],sep=""),intern=T),header=T,sep="\t")
          sumstat <- rbind(sumstat,x)
        }
        sumstat<-subset(sumstat, select=-c(X))

        TD_denom<-data.frame(sumstat[,grep("pi",colnames(sumstat))]-sumstat[,grep("theta_w",colnames(sumstat))])
        colnames(TD_denom)<-paste(colnames(sumstat)[grep("pi",colnames(sumstat))],
                                  colnames(sumstat)[grep("theta_w",colnames(sumstat))],sep="_")

        sumstat<-sumstat[,-grep("ZnS",colnames(sumstat))]
        sumstat<-sumstat[,-grep("thomson",colnames(sumstat))]
        sumstat<-cbind(sumstat, TD_denom)

        Mean<-colMeans(sumstat,na.rm = T)
        var<-apply(sumstat,2,var, na.rm=T)

        #kur<-apply(sumstat,2,kurtosis, na.rm=T)
        #skew<-apply(sumstat,2,skewness, na.rm=T)
        pp<-as.numeric(com[[nrow(model$loci)+1]][2,])
        param<-pp[1:(length(pp)-nrow(model$loci))]
        param<-c(param,mean(pp[(length(param)+1):length(pp)]))
        param<-c(param,sd(pp[(length(param)+1):length(pp)]))
        names(param)<-com[length(com)][[1]][1,1:length(param)]
        #simulations<-rbind(simulations,c(param,Mean,var))
        #file.remove(list.files(pattern = "out.txt"))
        simulations<-rbind(simulations,c(param,Mean,var))


      }

    }
    return(simulations)
  }
  parentls <- function()ls(envir=parent.frame())
  save(list=parentls(),file='.PM_objects.RData')

  thou<-0
  for(j in 1:nsim.blocks) {

    write(0,".log")

    for(c in 1:ncores){
      system(paste("Rscript .script_parallel.R",c), wait = F)
    }

    l<-0
    TIM1<-system.time(
    while(sum(as.numeric(l)) < ncores){
      Sys.sleep(1)
      l<-readLines(".log")
      })[3]

    file.remove(".log")

    simulations<-NULL
    TIM2<-system.time(
    for(t in 1:ncores){
      simulations<-rbind(simulations,read.table(paste(".",t,"SIMS_",output.name,".txt",sep=""), sep="\t"))
    })[3]
    TIM3<-system.time(
    write.table(simulations,file=paste("SIMS_",output.name,".txt",sep=""),quote=F,row.names = F,col.names = F, append=T,sep="\t")
    )[3]

  thou<-thou+block.size*ncores
  Total.time<-round(((((TIM1+TIM2+TIM3)*nsim.blocks)-((TIM1+TIM2+TIM3)*j))/60)/60,3)
  cat(paste("PipeMaster:: ",thou," (",round(((block.size*ncores)/(TIM1+TIM2+TIM3))*60*60)," sims/h) | ",Total.time," hours remaining",sep=""),"\n")
  #write.table(param,file=paste(output.name,"_par.txt",sep=""),quote=F,row.names = F,col.names = F, append=T,sep="\t")
  }

    }
